udp-client/src/nonblocking/udp_client.rs

```diff
@@ -0,0 +1,94 @@
+//! Simple UDP client that communicates with the given UDP port with UDP and provides
+//! an interface for sending transactions
+
+use {
+    async_trait::async_trait, core::iter::repeat, solana_sdk::transport::Result as TransportResult,
+    solana_streamer::nonblocking::sendmmsg::batch_send,
+    solana_tpu_client::nonblocking::tpu_connection::TpuConnection, std::net::SocketAddr,
+    tokio::net::UdpSocket,
+};
+
+pub struct UdpTpuConnection {
+    socket: UdpSocket,
+    addr: SocketAddr,
+}
+
+impl UdpTpuConnection {
+    pub fn new_from_addr(socket: std::net::UdpSocket, tpu_addr: SocketAddr) -> Self {
+        socket.set_nonblocking(true).unwrap();
+        let socket = UdpSocket::from_std(socket).unwrap();
+        Self {
+            socket,
+            addr: tpu_addr,
+        }
+    }
+}
+
+#[async_trait]
+impl TpuConnection for UdpTpuConnection {
+    fn tpu_addr(&self) -> &SocketAddr {
+        &self.addr
+    }
+
+    async fn send_wire_transaction<T>(&self, wire_transaction: T) -> TransportResult<()>
+    where
+        T: AsRef<[u8]> + Send + Sync,
+    {
+        self.socket
+            .send_to(wire_transaction.as_ref(), self.addr)
+            .await?;
```

Comment:
we can return the futures and let the caller decide when to `await`, no?

Reply:
Yes. Hmm, that's a small change we can make now, independent of this refactor.